### PR TITLE
v0.3.0 - Tags (F14 Versioning)

### PR DESCRIPTION
**Pull request by paulcjh**
# Pull request outline
This PR includes adds support for tag management on a remote compute service. The main utility of this is for versioning pipelines so that users can perform runs against a tag instead of a static `pipeline_id` or remote virtual pipeline.

This feature references feature [F14 Versioning](https://outline.mystic.ai/doc/f14-versioning-Um31eEVtdc) and linear project [Pipeline tagging](https://linear.app/mystic-ai/project/pipeline-tagging-b15244a2a7a0).

Also in this PR is a cleanup to the way CLI commands are structured in the code base. A follow up PR will show an additional clean up for better testing support on the CLI but it's out of the scope for this PR.

## Checklist:
- [x] Docs updated ([docs here](https://docs.pipeline.ai/v0.3.0/docs/pipeline-tags))
- [x] Tests written

~~- [ ] Check for breaking changes in Resource~~

- [x] Version bumped

Added:
- Pipeline tag schemas
- Tag management through CLI

Changed:
- Moved CLI commands into individual files

Removed:
_none_

## Related issues:
_none_

___

- _You can add a reference to an issue using hash followed by the issue number_
- _If a checklist item is to be ignored it must be struck through_

### CODE REVIEW

File: pipeline/console/tags.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        remote_service._get(f"/v2/pipeline-tags/by-name/{tag_name}")
+    )
+    return tag_information
+
+
+def _update_or_create_tag(source: str, target: str, sub_command: str) -> PipelineTagGet:
+    remote_service = PipelineCloud(verbose=False)
+    remote_service.authenticate()
+    if not tag_re_pattern.match(source):
+        _print("Source tag must match pattern 'pipeline:tag'", level="ERROR")
```
Comment (alexpearce) [inline pipeline/console/tags.py:33]: If we want to follow the `docker tag` CLI then `source` should be the 'origin' pipeline, i.e. the pipeline that the new tag `target` will point to.

```
$ docker tag create
"docker tag" requires exactly 2 arguments.
See 'docker tag --help'.

Usage:  docker tag SOURCE_IMAGE[:TAG] TARGET_IMAGE[:TAG]

Create a tag TARGET_IMAGE that refers to SOURCE_IMAGE
```

This `if not tag_re_pattern.match(source):` prevents the source (in the `docker tag` sense) being a pipeline ID.

File: pipeline/api/cloud.py [hunk 2]
```diff
@@ -332,7 +335,41 @@ def _post(self, endpoint, json_data):
...
+    def _delete(self, endpoint: str) -> dict:
+        self.raise_for_invalid_token()
+        headers = {
+            "Authorization": "Bearer %s" % self.token,
+            "Content-type": "application/json",
+        }
+
+        url = urllib.parse.urljoin(self.url, endpoint)
+        response = httpx.delete(url, headers=headers, timeout=self.timeout)
+        if response.status_code is not HTTPStatus.NO_CONTENT:
```
Comment (alexpearce) [inline pipeline/api/cloud.py:367]: `HTTPStatus.NO_CONTENT` is an enum whereas `response.status_code` is an integer, so comparison by `is` will fail.

```suggestion
        if response.status_code != HTTPStatus.NO_CONTENT:
```

File: tests/test_cli.py [hunk 3]
```diff
@@ -83,3 +96,26 @@ def test_runs_get(url, token, capsys, run_get, top_api_server):
...
+
+def test_tags_create(url: str, token: str, top_api_server: HTTPServer):
+    _set_testing_remote_compute_service(url=url, token=token)
+    with pytest.raises(SystemExit):
+        cli_main(["tags", "create", "bad_tag", "pipeline_id"])
+    assert cli_main(["tags", "create", "good:tag", "pipeline_id"]) == 0
+    assert cli_main(["tags", "create", "good:tag", "existing:tag"]) == 0
+
+    create_output = _update_or_create_tag("good:tag", "pipeline_id", "create")
+    print(create_output)
```
Comment (alexpearce) [inline tests/test_cli.py:114]:
```suggestion
```

**Review by alexpearce — CHANGES_REQUESTED** [review]
Good stuff here. Some mostly minor chages from me, with a suggestion on swapping the `pipeline tags create` order.

File: pipeline/console/tags.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        response = remote_service._get(
+            "/v2/pipeline-tags",
+            params=dict(
+                skip=skip,
+                limit=limit,
+                order_by="created_at:desc",
+            ),
+        )
+
+        paginated_results = Paginated[PipelineTagGet].parse_obj(response)
```
[the remark this answers is not in the document]
Comment (paulcjh) [reply pipeline/console/tags.py]: I lost the link to this, does the duplication still exist?

File: pipeline/console/tags.py (hunk 1 above)
Comment (paulcjh) [reply pipeline/console/tags.py]: done

File: pipeline/console/tags.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+from pipeline import PipelineCloud
+from pipeline.schemas.pagination import Paginated
+from pipeline.schemas.pipeline import (
+    PipelineTagCreate,
+    PipelineTagGet,
+    PipelineTagPatch,
+)
+from pipeline.util.logging import _print
+
+tag_re_pattern = re.compile(r"^[0-9a-zA-Z\-\_]+:[0-9a-zA-Z\-\_]+$", re.IGNORECASE)
```
Comment (alexpearce) [inline pipeline/console/tags.py:17]: I used a slightly more involved regex for the API service, but at least this wil catch obvious mistakes client side.

File: pipeline/console/tags.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            ]
+            for tag in tags
+        ],
+        headers=[
+            "ID",
+            "Name",
+            "Target",
+        ],
+        tablefmt="outline",
+    )
```
Comment (alexpearce) [inline pipeline/console/tags.py:112]: Would showing `created_at` here be useful for users?

**Review by alexpearce — APPROVED** [review]
Good stuff! Thanks a lot. I had a play around with this locally and seems to work as advertised 👍

**Pull request merged**